### PR TITLE
DAOS-6429 iosrv: Misc fixes in shutdown path

### PR DESCRIPTION
**Pull request by NiuYawei**
1. The srv TLS is accessible to all server modules, so it could
   be used by some module TLS init/fini functions. Let's register
   srv TLS in first place, so that it'll be first initialized and
   last finalized.

2. Pool module cleanup function should release all the ds_pool held
   from pool start.

3. Pool module TLS fini function shouldn't purge pool child cache,
   the pool child cache should have been purged in pool module
   cleanup function (which is called before TLS fini).

Features: rebuild

Signed-off-by: Niu Yawei <yawei.niu@intel.com>

### CODE REVIEW

File: src/iosrv/module.c [hunk 1]
```diff
@@ -253,6 +253,12 @@ dss_module_init_all(uint64_t *mod_facs)
 	uint64_t		 fac;
 	int			 rc = 0;
 
+	/*
+	 * first register global tls accessible to all modules,
+	 * it'll be first initialized and last finialized.
+	 */
+	dss_register_key(&daos_srv_modkey);
```
Comment (gnailzenh) [inline src/iosrv/module.c:260]: forgot to deregister?

Comment (NiuYawei) [reply src/iosrv/module.c:260]: actually, we didn't deregister any TLS key in current implementation. Deregister looks not quite necessary to me, but I think we may fix it in other cleanup patch.

**Review by gnailzenh — APPROVED** [review]

**Comment by NiuYawei** [comment]
The test result shows there is still other TLS fini order issue, pushed second commit to address it.

**Review by liuxuezhao — APPROVED** [review]

**Review by liuxuezhao — DISMISSED** [review]

File: src/pool/srv_target.c [hunk 2]
```diff
@@ -101,7 +101,8 @@ ds_pool_child_purge(struct pool_tls *tls)
 	d_list_for_each_entry_safe(child, n, &tls->dt_pool_list, spc_list) {
 		d_list_del_init(&child->spc_list);
 		ds_cont_child_stop_all(child);
-		D_ASSERTF(child->spc_ref == 1, DF_UUID": %d\n",
+		/* FIXME: Ref leaking in pool start, see ds_pool_stop_all() */
+		D_ASSERTF(child->spc_ref <= 2, DF_UUID": %d\n",
```
Comment (wangdi1) [inline src/pool/srv_target.c:105]: This will cover bug? probably just use D_WARN here?

**Review by byron-marohn — DISMISSED** [review]

**Comment by NiuYawei** [comment]
Didn't observe segfault or assert on server shutdown anymore, but "rebuild delete object" is still failing, I suppose that's known issue DAOS-6118.

**Comment by NiuYawei** [comment]
Discussed with Liwei, needs to revise the patch.

**Comment by NiuYawei** [comment]
sorry, I squashed patches and made a force.

File: src/pool/srv_pool.c [hunk 3]
```diff
@@ -1406,18 +1406,32 @@ ds_pool_start_all(void)
...
-	return ds_rsvc_stop_all(DS_RSVC_CLASS_POOL);
+	int	rc;
+
+	rc = ds_rsvc_stop_all(DS_RSVC_CLASS_POOL);
+	if (rc)
+		D_ERROR("failed to stop all pool svcs: "DF_RC"\n", DP_RC(rc));
+
+	rc = ds_mgmt_tgt_pool_iterate(stop_one, NULL /* arg */);
+	if (rc != 0)
+		D_ERROR("failed to stop all pools: "DF_RC"\n", DP_RC(rc));
```
Comment (liw) [inline src/pool/srv_pool.c:1432]: An improvement for sure. This, however, scans the local file system for local pools. I think the best approach is still to scan the ds_pool LRU hash table.

**Review by liw — DISMISSED** [review]
Thanks, Niu.

**Comment by NiuYawei** [comment]
Didn't observe segfault or assert on shutdown in last two round of CI testing (with Feature: rebuild set), but hit intermittent rebuild 27 failure on last run. I believe the patch has already fixed the issue on shutdown path, let me rebase the patch to try test again.

**Comment by NiuYawei** [comment]
Fixed few defects in the patch, All CI tests with "rebuild" features passed now. Please re-review it. Thanks!

**Review by liw — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Pull request merged**